src/language/rule/rule.ts

```diff
@@ -265,3 +278,37 @@ export class RuleFailure {
         return new RuleFailurePosition(position, lineAndCharacter);
     }
 }
+
+export class WalkContext<T> {
+    private limit: number;
+    private failures: RuleFailure[];
+
+    constructor(public readonly sourceFile: ts.SourceFile, public readonly ruleName: string, public readonly options: T) {
+        this.failures = [];
+        this.limit = sourceFile.getFullWidth();
+    }
+
+    public getFailures() {
+        return this.failures;
+    }
+
+    /** Add a failure with any arbitrary span. Prefer `addFailureAtNode` if possible. */
+    public addFailureAt(start: number, width: number, failure: string, fix?: Fix) {
+        this.addFailure(start, start + width, failure, fix);
+    }
+
+    public addFailure(start: number, end: number, failure: string, fix?: Fix) {
```

Comment:
Assuming all rules will eventually be migrated over to the new style, I'm in agreement. The type system will find incorrect usages though, and the new naming is nicer.

If some rules are going to stay as extensions of `RuleWalker` though, then it'll get confusing to have the different methods existing for perpetuity.